Working toward fixing #598

PhilBax opened the pull request.

PhilBax commented:
@googlebot I signed it!

kqyang commented on packager/media/formats/mp2t/adts_header.cc:
```diff
@@ -63,7 +63,7 @@ bool AdtsHeader::Parse(const uint8_t* adts_frame, size_t adts_frame_size) {
   RCHECK(frame.SkipBits(1));
   RCHECK(frame.ReadBits(3, &channel_configuration_));
-  RCHECK((channel_configuration_ > 0) &&
+  RCHECK((channel_configuration_ >= 0) &&
```
This condition can be removed as |channel_configuration_| is unsigned.

PhilBax replied:
Oh! So it is!

kqyang commented on packager/media/formats/mp2t/adts_header.cc:
```diff
@@ -109,7 +117,7 @@ uint32_t AdtsHeader::GetSamplingFrequency() const {
 
 uint8_t AdtsHeader::GetNumChannels() const {
-  DCHECK_GT(channel_configuration_, 0);
+  DCHECK_GE(channel_configuration_, 0);
```
The line can be removed as |channel_configuration_| is unsigned.

kqyang commented on packager/media/formats/mp2t/audio_header.h:
```diff
@@ -50,6 +50,8 @@ class AudioHeader {
   /// @return the size of frame (header + payload).
   virtual size_t GetFrameSize() const = 0;
 
+  virtual size_t GetFrameSizeWithoutParsing(const uint8_t* data, size_t num_bytes) const {return 0;}
```
Can you add a comment like the functions above? Also define it as pure virtual.

You'll also need to implement the similar function in ac3_header.h.

PhilBax replied:
Yeah, this was just a quick-and-dirty implementation to get it to build. I haven't looked at the ac3 side of things. I might ping you if it's not immediately obvious to me how to implement this function for AC3 audio.

PhilBax replied:
Ah! Okay, I get was the adts version was doing. Should be able to handle the ac3 version, I think.

PhilBax replied:
Since GetFrameSize() actually has to do some logic, would it be bad form to have GetFrameSizeWithoutParsing() actually set fscod_ and frmsizecod_ and then call GetFrameSize() (though, I guess that would make the function non-const)?

Would it be best to have both GetFrameSizes call a common function to do the actual frame size logic?

Or is the AC3 standard fixed enough that duplicating the small amount of logic involved is fine by y'all?

PhilBax replied:
Like, does this work for you?

```
/// Calculate the size of the frame from the sample rate code and the 
/// frame size code.
/// @return the size of the frame (header + payload).
size_t CalcFrameSize(uint8_t fscod, uint8_t frmsizecod) {
  const size_t kNumFscode = arraysize(kAc3SampleRateTable);
  DCHECK_LT(fscod, kNumFscode);
  DCHECK_LT(frmsizecod, arraysize(kFrameSizeCodeTable));
  // The order of frequencies are reversed in |kFrameSizeCodeTable| compared to
  // |kAc3SampleRateTable|.
  const int index = kNumFscode - 1 - fscod;
  return kFrameSizeCodeTable[frmsizecod][index] * 2;
}

size_t Ac3Header::GetFrameSize() const {
  return CalcFrameSize(fscod_, frmsizecod_);
}

size_t Ac3Header::GetFrameSizeWithoutParsing(const uint8_t* data, size_t num_bytes) const {
  uint8_t fscod = data[5] & 0xc0;
  uint8_t frmsizecod = data[5] & 0x3f;
  return CalcFrameSize(fscod, frmsizecod);
}
```

kqyang replied:
Yes, that looks very good to me!

> uint8_t fscod = data[5] & 0xc0;

Should be data[4] instead of data[5].

Also, there should be check of 'num_bytes' to make sure that it is >4.

PhilBax replied:
Good catch! I hadn't actually tested it yet. 'preciate it!

kqyang commented on packager/media/formats/mp2t/es_parser_audio.cc:
```diff
@@ -56,20 +56,39 @@ static bool LookForSyncWord(const uint8_t* raw_es,
     const int kSyncWordSize = 2;
-    if (remaining_size >= audio_header->GetFrameSize() + kSyncWordSize &&
-        !audio_header->IsSyncWord(&cur_buf[audio_header->GetFrameSize()])) {
+    const size_t frame_size = audio_header->GetFrameSizeWithoutParsing(cur_buf, audio_header->GetMinFrameSize()); /* Get Frame size without parsing the frame */
```
audio_header->GetMinFrameSize() should be replaced with remaining_size.

kqyang commented on packager/media/formats/mp2t/es_parser_audio.cc:
```diff
@@ -56,20 +56,39 @@ static bool LookForSyncWord(const uint8_t* raw_es,
+    // if (remaining_size >= audio_header->GetFrameSize() + kSyncWordSize &&
+    //     !audio_header->IsSyncWord(&cur_buf[audio_header->GetFrameSize()])) {
+    //   continue;
+    // }
```
Remove these commented codes.

PhilBax commented:
Should I squash the pull request down to a single commit? Do you do squash-merges? Do you prefer full-history?

PhilBax commented:
> Can you run "git clang-format " to format the code properly?

TIL! :smile:

kqyang commented:
> Should I squash the pull request down to a single commit? Do you do squash-merges? Do you prefer full-history?

Yes, I will do "squash and merge", so you don't need to squash the pull request into a single commit.

Btw, if you don't have 'clang-format' installed already, you can follow this instruction to install it: https://github.com/google/shaka-packager/blob/master/packager/tools/git/check_formatting.py#L16.

Thanks again for your help!

PhilBax commented:
I did exactly that! I'd seen clang-format before, but didn't know about the git integration. Really neat tool!

PhilBax commented:
For the unit tests... do you want *new* unit tests? Or should I just add an EXPECT_EQ to the existing parse tests, but before the parse actually happens? Technically the new function is "parsing" a small amount of info out of the header, and it would be nice to just go ahead and use the setup for the existing three ac3 tests. But I fully understand making it separate as well.

If new tests, should I add one to match each of the ac3 "parse" tests? Or just a single one for ac3 and one for adts?

kqyang commented:
> should I just add an EXPECT_EQ to the existing parse tests, but before the parse actually happens?

SGTM!

PhilBax commented:
K, should be good-to-go I think. Fixed a logic issue with the AC3 parsing in the code I pasted in the above comment, and added some unit tests which are now passing!

kqyang commented on packager/media/formats/mp2t/ac3_header.cc:
```diff
@@ -103,14 +103,28 @@ size_t Ac3Header::GetHeaderSize() const {
-  return kFrameSizeCodeTable[frmsizecod_][index] * 2;
+  const int index = kNumFscode - 1 - fscod;
+  return kFrameSizeCodeTable[frmsizecod][index] * 2;
+}
```
Can you move it to the anonymous namespace above, i.e. after line 42?

PhilBax replied:
Sure can!

kqyang commented on packager/media/formats/mp2t/ac3_header.cc:
```diff
@@ -103,14 +103,28 @@ size_t Ac3Header::GetHeaderSize() const {
-size_t Ac3Header::GetFrameSize() const {
+/// Calculate the size of the frame from the sample rate code and the 
+/// frame size code.
+/// @return the size of the frame (header + payload).
```
nit: Use normal comment style for internal functions, e.g.

// Calculate the size of the frame from the sample rate code and the 
// frame size code.

PhilBax replied:
Ah! I wondered. Should've checked the style guide.

kqyang commented on packager/media/formats/mp2t/ac3_header.cc:
```diff
@@ -103,14 +103,28 @@ size_t Ac3Header::GetHeaderSize() const {
+  return CalcFrameSize(fscod_, frmsizecod_);
+}
+
+size_t Ac3Header::GetFrameSizeWithoutParsing(const uint8_t* data, size_t num_bytes) const {
```
This is longer than 80 characters. Have you run "git clang-format"?

PhilBax replied:
I did, but I committed this before running it. Apparently git clang-format only runs on the lines you're actively submitting? I presume there must be a flag to run on the whole file. Will look into it!

kqyang replied:
You can run `git clang-format <base_commit>`

For example, if `HEAD^` is the base commit, you can run `git clang-format HEAD^`

PhilBax replied:
Gotcha! Done!

kqyang commented on packager/media/formats/mp2t/es_parser_audio.cc:
```diff
@@ -56,20 +56,26 @@ static bool LookForSyncWord(const uint8_t* raw_es,
+     if (!audio_header->IsSyncWord(cur_buf))
       continue;
 
     // Check whether there is another frame |size| apart from the current one.
```
Can you move this comment before line 71?

kqyang commented on packager/media/formats/mp2t/es_parser_audio.cc:
```diff
@@ -56,20 +56,26 @@ static bool LookForSyncWord(const uint8_t* raw_es,
       continue;
 
     // Check whether there is another frame |size| apart from the current one.
     const size_t remaining_size = static_cast<size_t>(raw_es_size - offset);
```
Thinking about it again, I think there should be another condition here:

```c
if (remaining_size < audio_header->GetMinFrameSize()) {
  // Not a full frame: will resume when we have more data.
  return false;
}
```

kqyang commented on packager/media/formats/mp2t/es_parser_audio.cc:
```diff
@@ -56,20 +56,26 @@ static bool LookForSyncWord(const uint8_t* raw_es,
+    const size_t frame_size =
+        audio_header->GetFrameSizeWithoutParsing(cur_buf, remaining_size);
+    if (frame_size < audio_header->GetMinFrameSize())
+      continue;
```
Can you add a comment here:

```c
if (frame_size < audio_header->GetMinFrameSize()) {
  // Too short to be a valid frame.
  continue;
}
```

PhilBax replied:
Sure thing!

kqyang commented on packager/media/formats/mp2t/es_parser_audio.cc:
```diff
@@ -56,20 +56,26 @@ static bool LookForSyncWord(const uint8_t* raw_es,
+    if (frame_size < audio_header->GetMinFrameSize())
+      continue;
+    if (remaining_size < frame_size)
+      return false;
```
Similarly,

```c
    if (remaining_size < frame_size) {
      // Not a full frame: will resume when we have more data.
      return false;
    }
```

kqyang commented on packager/media/formats/mp2t/es_parser_audio.cc:
```diff
@@ -59,17 +59,29 @@ static bool LookForSyncWord(const uint8_t* raw_es,
+      return false;
+    if (remaining_size < audio_header->GetMinFrameSize()) {
+      // Not a full frame: will resume when we have more data.
+      return false;
```
Just realized that this will never occur as 

remaining_size = raw_es_size - offset >= raw_es_size - (raw_es_size - audio_header->GetMinFrameSize()) = audio_header->GetMinFrameSize(). 

Can you remove the three lines from 72 - 74?

Sorry about the previous suggestion earlier.

kqyang commented on packager/media/formats/mp2t/adts_header.cc:
```diff
@@ -89,6 +88,14 @@ size_t AdtsHeader::GetFrameSize() const {
+size_t AdtsHeader::GetFrameSizeWithoutParsing(const uint8_t* data,
+                                              size_t num_bytes) const {
+  if (num_bytes < 6)
+    return 0;
```
For consistency with Ac3Header, replace it with DCHECK_GT(num_bytes, 5)?

kqyang commented on packager/media/formats/mp2t/ac3_header.cc:
```diff
@@ -104,13 +117,15 @@ size_t Ac3Header::GetHeaderSize() const {
+size_t Ac3Header::GetFrameSizeWithoutParsing(const uint8_t* data,
+                                             size_t num_bytes) const {
+  DCHECK_GT(num_bytes, static_cast<size_t>(4));
+  uint8_t fscod = (data[4] & 0xc0) >> 6;
```
nit: can be replaced with

uint8_t fscod = data[4] >> 6;

PhilBax replied:
Ah! Fair enough!

kqyang commented on packager/media/formats/mp2t/adts_header.cc:
```diff
@@ -89,6 +88,13 @@ size_t AdtsHeader::GetFrameSize() const {
 
+size_t AdtsHeader::GetFrameSizeWithoutParsing(const uint8_t* data,
+                                              size_t num_bytes) const {
+  DCHECK_GT(num_bytes, 5)
```
Missing ";" in the end. 

It causing building failures: https://ci.appveyor.com/project/shaka/shaka-packager/builds/27774035

PhilBax replied:
Oi! My lack of sleep is showing. :smile:

PhilBax replied:
Also, I have separate folders for the "official" shaka build, and my fork (which lacks the build-related files). I should've taken the time to overlay my fork on top of the folder with the build files.

kqyang commented on packager/media/formats/mp2t/adts_header.cc:
```diff
@@ -89,6 +88,13 @@ size_t AdtsHeader::GetFrameSize() const {
 
+size_t AdtsHeader::GetFrameSizeWithoutParsing(const uint8_t* data,
+                                              size_t num_bytes) const {
+  DCHECK_GT(num_bytes, 5);
```
Ah, there are still failures on Windows: https://travis-ci.org/google/shaka-packager/builds/591706925?utm_source=github_status&utm_medium=notification

You can probably fix it by either

```c++
DCHECK_GT(num_bytes, 5u);
```

or 

```C++
DCHECK_GT(num_bytes, static_cast<size_t>(5));
```

kqyang commented:
@PhilBax The CL is now merged. Thanks for improving Shaka Packager!

PhilBax commented:
Thank you for all the help!!
